// jmap.js/src/ZoneKit/ZoneCollection.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', './Zone.js'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, require('./Zone.js'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, global.Zone);
    global.ZoneCollection = mod.exports;
  }
})(this, function (module, Zone) {
  'use strict';

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var ZoneCollection = function (_global$jmap$core$Bas) {
    _inherits(ZoneCollection, _global$jmap$core$Bas);

    /**
     * Create a collection of Devices.
     */
    function ZoneCollection() {
      _classCallCheck(this, ZoneCollection);

      return _possibleConstructorReturn(this, (ZoneCollection.__proto__ || Object.getPrototypeOf(ZoneCollection)).call(this, Zone));
    }

    /**
     * Returns a boolean for whather or not argument is constructed as an Device object
     * @param {Object} item - Item to evaluate
     * @return {Boolean} Boolean based on evaluation result
     */


    _createClass(ZoneCollection, [{
      key: 'isZone',
      value: function isZone(item) {
        return this.isCollectionItemType(item);
      }
    }, {
      key: 'getByName',
      value: function getByName(name) {
        return this.filter(function (a) {
          return a.name === name.toLowerCase();
        });
      }
    }, {
      key: 'getByWaypoint',
      value: function getByWaypoint() {
        var waypoint = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

        return this.filter(function (z) {
          return z.getLocations().find(function (l) {
            return l.waypointIds.indexOf(waypoint.id) > -1;
          });
        });
      }
    }, {
      key: 'getByMap',
      value: function getByMap() {
        var map = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

        return this.filter(function (z) {
          return z.getLocations().find(function (l) {
            return l.mapId === map.id;
          });
        });
      }
    }]);

    return ZoneCollection;
  }(global.jmap.core.BaseCollection);

  module.exports = ZoneCollection;
});
