// @jibestream-dev/jmap-mapui-kit/karma.conf.js
module.exports = config => {
  config.set({
    client: {
      mocha: {
        timeout: 30000,
      },
    },
    frameworks: ['mocha', 'chai', 'browserify'],
    preprocessors: {
      'test/**/*.js': ['browserify'],
      'src/**/*.js': ['browserify', 'sourcemap'],
    },
    plugins: [
      'karma-chrome-launcher',
      'karma-browserify',
      'karma-mocha',
      'karma-chai',
      'karma-coverage',
      'karma-sourcemap-loader',
    ],
    files: [
      { pattern: 'src/_require.js' },
      { pattern: 'src/**/*.js' },
      { pattern: 'test/**/*.js' },
    ],
    browserify: {
      debug: true,
      transform: [
        'babelify',
        ['browserify-istanbul',
          {
            instrumenterConfig: {
              embedSource: true, // this is important for HTML reports
            },
            ignore: [
              '**/test/**/*.js',
            ],
          },
        ],
      ],
      extensions: ['.js'],
    },
    reporters: ['progress', 'coverage'],
    port: 9876, // karma web server port
    colors: true,
    logLevel: config.LOG_INFO,
    browsers: ['ChromeHeadless'],
    autoWatch: false,
    singleRun: true, // Karma captures browsers, runs the tests and exits
    concurrency: Infinity,
    // optionally, configure the reporter
    // coverageReporter: {
    //   reporters: [
    //     { type: 'text' },
    //     { type: 'html', dir: 'coverage/' },
    //   ],
    //   check: {
    //     global: {
    //       statements: 0,
    //     },
    //     each: {
    //       statements: 0,
    //     },
    //   },
    // },
  })
}
